src/hooks/useSignChallenge.ts

```diff
@@ -0,0 +1,123 @@
+
+  const storageKey = `${storageKeys.SIWE_SIGNATURE_KEY_PREFIX}${address}`;
+
+  const checkStoredSignature = useCallback((): SiweSignatureData | null => {
+    if (!address) return null;
+
+    try {
+      const stored = localStorage.getItem(storageKey);
+      if (!stored) return null;
+
+      const data: SiweSignatureData = JSON.parse(stored);
+      return new Date(data.expirationDate) > new Date() ? data : null;
+    } catch {
+      localStorage.removeItem(storageKey);
+      return null;
+    }
+  }, [address, storageKey]);
+
+  const signMessage = useCallback((): Promise<SiweSignatureData> => {
+    return new Promise((resolve, reject) => {
+      signPromiseRef.current = { resolve, reject };
+      setRequiresSign(true);
+    });
+  }, [setRequiresSign]);
+
+  const handleSign = useCallback(async () => {
+    if (!address || !signPromiseRef.current) return;
+
+    try {
+      const response = await fetch(`${SIGNING_SERVICE_URL}/v1/siwe/create`, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ walletAddress: address }),
+      });
+
+      if (!response.ok) throw new Error('Failed to create message');
+      const { siweMessage, nonce } = await response.json();
+
+      const message = new SiweMessage(siweMessage);
+      const signature = await signMessageAsync({ message: siweMessage });
+
+      const validationResponse = await fetch(`${SIGNING_SERVICE_URL}/v1/siwe/validate`, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ nonce, signature }),
+      });
+
+      if (!validationResponse.ok) throw new Error('Failed to validate signature');
+
+      const signatureData: SiweSignatureData = {
+        nonce,
+        signature,
+        expirationDate: message.expirationTime!,
+      };
+
+      localStorage.setItem(storageKey, JSON.stringify(signatureData));
+      signPromiseRef.current.resolve(signatureData);
+    } catch (error) {
+      signPromiseRef.current.reject(new Error('Signing failed'));
+    } finally {
+      setRequiresSign(false);
+      signPromiseRef.current = null;
+    }
+  }, [address, signMessageAsync, storageKey]);
+
+  // Handler for modal cancellation
+  const handleCancel = useCallback(() => {
+    if (signPromiseRef.current) {
+      signPromiseRef.current.reject(new Error('User cancelled'));
+      signPromiseRef.current = null;
+    }
+    setRequiresSign(false);
+  }, []);
+
+  const checkAndWaitForSignature = useCallback(async (): Promise<SiweSignatureData> => {
+    const stored = checkStoredSignature();
+    if (stored) return stored;
+    return signMessage();
+  }, [checkStoredSignature, signMessage]);
+
+  const forceRefreshAndWaitForSignature = useCallback(async (): Promise<SiweSignatureData> => {
+    localStorage.removeItem(storageKey);
+    return signMessage();
+  }, [storageKey, signMessage]);
+
+  // ask for signature on address change and on init
```

Comment:
I found this a bit invasive and annoying. But could be enabled. Perhaps with a timer.

Reply:
I think we should enable it so that users are immediately requested to log in after connecting the wallet. Makes sense to combine these two events so that the user doesn't need to sign thrice later.

Reply:
I'll enable it so we can test it. I worry a bit that is invasive in the sense that if the user only wants to see quotes or offramp `EURC`, it's not needed and may scare some users. 

On the other hand it's even more spaced from the other 2 signatures so "feels" better.